### PR TITLE
Use repositories

### PR DESCRIPTION
**Pull request by koenvo**

### CODE REVIEW

File: statsbombapi/models/data.py [hunk 1]
```diff
@@ -244,7 +244,7 @@ def __post_init__(self):
 @dataclasses.dataclass(frozen=True)
 class Lineup:
     team_id: int
-    team_name: int
+    team_name: str
```
Comment (Torvaney) [inline statsbombapi/models/data.py:247]: Good catch! It's a shame the json parser doesn't throw an error here

File: test/test_json.py [hunk 2]
```diff
@@ -29,15 +31,15 @@ def test_extract_unit():
     competition_season = sb_json.CompetitionSeason(
         competition_id=1,
         competition_name='ok',
-        competition_gender='female',
+        competition_gender=sb_json.Gender.FEMALE,
```
Comment (Torvaney) [inline test/test_json.py:34]: :+1:

File: statsbombapi/client.py [hunk 3]
```diff
@@ -1,122 +1,55 @@
...
+from .infrastructure import (
+    ReadOnlyRepositoryInterface, CachingRepositoryProxy,
+    get_github_repository,
+    get_local_repository,
+    get_statsbomb_services_repository
+)
+
+
+class APIClient:
+    def __init__(self, repository: ReadOnlyRepositoryInterface, caching_dir=None):
```
Comment (Torvaney) [inline statsbombapi/client.py:13]: What do you think about adding a note to the readme highlighting the ability to cache? At the moment, it's not clear to users that this is supplied functionality.

**Review by Torvaney — COMMENTED** [review]
Hey Koen,

Thanks for the contribution and for your patience in waiting for me to get around to this. Overall, this looks really cool and I really like the extended functionality - the addition of local data sources was something I hadn't thought much about, but could be really useful for ETL pipelines. 

I have a few outstanding questions on the changes here; although I'm not that experienced with Design Patterns and OOP, so you may have to bear with me if some of my questions are a little basic!

The two main questions are about the overall structure of the library. The first is about the use of the APIClient class vs Repositories, and the second is about the motivation behind the saving features added through the Adapter and Serialiser classes. I also have a few other minor questions and comments that are written below.

**APIClient vs Repository**

As I see it, the `APIClient` class is the product of a `Repository` and whether or not to use a cache (`caching_dir`). As a result, each of the APIClient's methods are simple forwarding methods onto the Repository. This has me wondering -  what value does the extra layer of indirection from adding a new class give us? Could we simplify things by exposing the repository instances directly? After all, the signatures of the `get_*_repository` and `get_*_client` functions are virtually identical.through

The caching functionality could still be exposed through an argument to any `get_*_client` (or whatever the equivalent ends up being) function, or by exposing the `CachingRepositoryProxy` class to the user directly.

**Saving API data**

The caching functionality is a nice, sensible feature that has obvious applications. However, it comes at the cost of some additional complexity - namely, the addition of a serialiser class, and repositories as the product of an adapter and a serialiser. Given the relatively limited scope of what we're doing here, I can't help but feel like we could make the implementation a bit more concrete - trading off some of the generality for simpler, less flexible code for this specific use-case.

Speaking in more concrete terms, we basically have 2 types of repository:

* JSON-getter + DataclassesJsonSerializer - Essentially, an Adapter that gets JSON from someplace, which has each route parsed with the functions in `statsbombapi.models.parse`.
* The CachingRepositoryProxy - the only repository that uses the write functionality

I am not convinced we are benefiting from the additional abstraction and generality from making the caching fit the same structure as the main 3 repositories (github, local, statsbomb), especially when those 3 are so similar.

In addition - what is the reasoning behind caching with pickle? As I see it, we could cache the JSON itself, which would have a few advantages:

* Improved security - Pickle's issues with arbitrary code execution are well documented and probably not something we'd like to expose users to.
* Compatibility - JSON will be compatible across different versions of this package (and/or across different packages, programming languages, etc.)

Furthermore, if we're only reading and writing to/from JSON, then there's only one serialisation format needed, and a single serialiser can be shared across all repositories. This strikes me as an opportunity to make the whole thing a little less abstract. If all repositories are using the same serialiser, then we might not need the abstraction of a serialiser at all!

I believe that making this implementation simpler (and less abstract) would have benefits in the maintainability and approachability of the code (I include approachability because many sports analytics enthusiasts are not expert software engineers, including me!). It may be that we need to jump back up to this level of abstraction in the future, when there are users with problems they cannot solve without it. However, I am wary of starting out at this point, before the package has any users.

What do you think? I am aware that I may not have fully grasped all of the reasoning behind these changes, so please do let me know if what I have written makes sense and if there are any points which don't add up.

**Other questions/comments**

* There is a bunch of new code written, but no new tests. Is this an issue? A lot of the new code isn't adding new _functionality_ so I can see why tests might not be adding a huge amount of value. What do you think?
* Could you update the package version to `0.1.0`, please? Seems worth noting that this is a significant change in the features and contents of the package.
* Could you add yourself as a co-author, please? This is a significant contribution, which should be reflected in the package metadata.

Thanks again for opening this PR - I have learned a lot from reading through your changes and seeing the patterns you've applied!

**Comment by Torvaney** [comment]
Also - I updated the license to use GPLv3 - hopefully that doesn't change your desire to contribute but I thought I should give you the heads up anyway :)

**Comment by koenvo** [comment]
Hi Ben,

Thanks for looking in this pull request!

After reading your feedback I must admit it looks a bit like over-engineering. I tried to not-overdo abstractions but I might have failed there :-)

Lets look into your questions.

**APIClient vs Repository** 
You are totally right the APIClient is redundant in the implementation with repositories. The main reason I kept it was to keep it simple for end-users. They don't have to think about the repositories underneath it. But I think by trying to make it simple it became more complex..

I believe the main interface could be the repository returned by `get_github_repository` etc methods in `repositories/ __init__.py`. The use of this abstractions make it super easy to replace an adapter with a database connector to store/fetch everything from a database.

**Saving API data**
The additional functionality of being able to write it to disk was added with an ETL process in mind. When running a script over and over again it's a waste of time and server capacity to fetch the data over and over. Al though in the statsbombpy package they  provide a method (https://github.com/statsbomb/statsbombpy/blob/master/statsbombpy/sb.py#L87-L117) to 'scrape' their api, without caching so the server capacity shouldn't be that much of a problem on their side.

I would think of a ETL process like:
1. Fetch data from statsbomb and store it locally / database
2. Fetch data from local disk / database and proces it, write results somewhere

And you are right about the caching proxy. It should be made available through something like a  "utils" module, and documented together with an example of how to use it. Or maybe even let people figure out how to roll it on their own.

The only reason for the pickle is performance. Loading the events using the current json implementation takes 12seconds while loading it from pickle format it takes 0.1second. Oh.. and that implementing the write methods was super easy. Security shouldn't be a problem because the library also writes the pickle files. You should never unpickle data from an unknown source. And indeed... versioning is a problem.

**Simple vs complex**
I introduced some additional layers which require some thinking about the code. I'm aware of that. The most important reason is that on the long term it makes code easier to think about. Every component of the code has a single responsibility and it should be explainable. In this case:
1. Repository is the main interface to fetching/storing data
2. Serializer is responsible for (de)serializing from data->object and visa versa
3. Adapter is used for reading/writing the data in serialised form.

Swapping out an Serializer / Adapter is pretty easy.  Think about reading/writing to a database:

```python

database_repo = Repository(
  adapter=DBAdapter(url="postgres://......@...."),
  serializer=DataclassesJsonSerializer()
)
```

**Other questions**
Yes there should be tests for this! And I will change version and maintainers, thanks!

Looking forward to your comment on my story :-)

**Comment by Torvaney** [comment]
Hey Koen

Sorry again for the long wait. Let's just say I am enjoying the slow and steady approach.. 😬 

Having reflected a bit more, and experimented a bit more with using the package, I can see the value of this architecture a bit better.

> APIClient vs Repository

You seem to agree with some of my points, here. What do you think the best course of action for this is in this PR? Simplify or stick with what's in the PR now?

> Saving API data

Completely agree on both hitting the API endpoint over and over, and also on speed with pickling. This is a really nice addition in real use.

> And you are right about the caching proxy. It should be made available through something like a "utils" module, and documented together with an example of how to use it. Or maybe even let people figure out how to roll it on their own.

Both these options sound good to me. What do you think the best one is for now?

Thanks again for your patience. I am happy with the changes if we can agree on a couple of the remaining action points.

**Comment by koenvo** [comment]
Let me update my PR and simplify by replacing ApiClient with repository, but also keeping the api as it is right now. I'll also move the caching to some util function. GPLv3 isn't a problem btw.

**Comment by Torvaney** [comment]
> Let me update my PR and simplify by replacing ApiClient with repository, but also keeping the api as it is right now. I'll also move the caching to some util function. GPLv3 isn't a problem btw.

That’s awesome, thanks mate

**Comment by Torvaney** [comment]
Hey @koenvo, I've made some changes to the way this package is structured. These changes build on the proposed changes from this PR, but aren't as far reaching. The feedback proposed here was instrumental in the redesign, so I just wanted to say thank you for the time taken in the above PR.

In light of the [changes](https://github.com/Torvaney/statsbombapi/commit/a6c9b20e279030f6ea416906dfe75347ea77a8c3), I'm going to close this PR. Happy to re-open or discuss further improvements, though.

**Pull request closed without merging**